### ISSUE TITLE
Half-Precision Optimizer

### ISSUE THREAD

**Issue by gupta-abhay** [google-research/vision_transformer#1]
Hello,

Thanks for the code. 

Looking at the `momentum_hp.py` file, at line 22, you comment that the optimizer `stores state using half-precision`. Just to clarify, are you using a half-precision optimizer similar to the [Jukebox](https://arxiv.org/abs/2005.00341) paper for training?

**Comment by andsteing** [comment]
You can see at [line 39](https://github.com/google-research/vision_transformer/blob/master/vit_jax/momentum_hp.py#L39) that only the momentum is kept at half precision (`bfloat16` to be precise). All other arrays are kept as `float32`.

You can also check this in the [Colab](https://colab.sandbox.google.com/github/google-research/vision_transformer/blob/master/vit_jax.ipynb) by inspecting the optimizer:

```
jax.tree_map(
    lambda t: t.dtype if isinstance(t, jax.numpy.ndarray) else None,
    flax.serialization.to_state_dict(opt),
)
```

As for in which precision the computations are done, please refer to JAX's [Type promotion semantics](https://jax.readthedocs.io/en/latest/type_promotion.html).

I haven't looked at the Jukebox code, but they state in their paper:

> ... fully half-precision training, i.e. storing the model weights, gradients and the optimizer states in half precision and performing computations in half precision as well.
